br/pkg/lightning/restore/check_info.go

```diff
@@ -399,535 +180,64 @@ func (rc *Controller) estimateSourceData(ctx context.Context) (int64, int64, boo
-			return nil
-		}
-
-		for i := 0; i < len(rows[0]); i++ {
-			if rows[0][i].GetString() != rows[1][i].GetString() {
-				return nil
-			}
-		}
-	}
-
-	// check if some fields are unique and not ignored
-	// if at least one field appears in a unique key, we can sure there is something wrong,
-	// they should be either the header line or the data is duplicated.
-	tableInfo := dbInfos[tableMeta.DB].Tables[tableMeta.Name]
-	tableFields := make(map[string]struct{})
-	uniqueIdxFields := make(map[string]struct{})
-	ignoreColumns, err := rc.cfg.Mydumper.IgnoreColumns.GetIgnoreColumns(tableMeta.DB, tableMeta.Name, rc.cfg.Mydumper.CaseSensitive)
+	result, err := theChecker.Check(ctx)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	ignoreColsSet := make(map[string]struct{})
-	for _, col := range ignoreColumns.Columns {
-		ignoreColsSet[col] = struct{}{}
-	}
-	for _, idx := range tableInfo.Core.Indices {
-		if !idx.Unique && !idx.Primary {
-			continue
-		}
-		for _, col := range idx.Columns {
-			if _, ok := ignoreColsSet[col.Name.L]; !ok {
-				uniqueIdxFields[col.Name.L] = struct{}{}
-			}
-		}
-	}
-	for _, f := range tableInfo.Core.Columns {
-		tableFields[f.Name.L] = struct{}{}
-	}
-	if common.TableHasAutoRowID(tableInfo.Core) {
-		tableFields[model.ExtraHandleName.L] = struct{}{}
-	}
-	hasUniqueField := false
-	for _, d := range rows[0] {
-		val := strings.ToLower(d.GetString())
-		if _, ok := tableFields[val]; !ok {
-			return nil
-		}
-		if _, ok := uniqueIdxFields[val]; ok {
-			hasUniqueField = true
-			break
-		}
-	}
-
-	msg := fmt.Sprintf("source csv files contains header row but `mydumper.csv.header` is false, checked table is `%s`.`%s`",
-		tableMeta.DB, tableMeta.Name)
-	level := Warn
-	if hasUniqueField && len(rows) > 1 {
-		level = Critical
-	} else if !checkFieldCompatibility(tableInfo.Core, ignoreColsSet, rows[0], log.FromContext(ctx)) {
-		// if there are only 1 csv file or there is not unique key, try to check if all columns are compatible with string value
-		level = Critical
+	if result != nil {
+		rc.checkTemplate.Collect(result.Severity, result.Passed, result.Message)
 	}
-	rc.checkTemplate.Collect(level, false, msg)
-
 	return nil
 }
 
-func checkFieldCompatibility(
-	tbl *model.TableInfo,
-	ignoreCols map[string]struct{},
-	values []types.Datum,
-	logger log.Logger,
-) bool {
-	se := kv.NewSession(&kv.SessionOptions{
-		SQLMode: mysql.ModeStrictTransTables,
-	}, logger)
-	for i, col := range tbl.Columns {
-		// do not check ignored columns
-		if _, ok := ignoreCols[col.Name.L]; ok {
-			continue
-		}
-		if i >= len(values) {
-			break
-		}
-		_, err := table.CastValue(se, values[i], col, true, false)
-		if err != nil {
-			logger.Error("field value is not consistent with column type", zap.String("value", values[i].GetString()),
-				zap.Any("column_info", col), zap.Error(err))
-			return false
-		}
-	}
-
-	return true
-}
-
 func (rc *Controller) checkTableEmpty(ctx context.Context) error {
```

Comment:
Main logic is extracted as `(*tableEmptyCheckItem).Check()`, in `br/pkg/lightning/restore/precheck_impl.go`, L1071